compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerCoalescedDMAToGlobalLoads.cpp

```diff
@@ -0,0 +1,383 @@
+//     %thread_index = affine.delinearize_index(%subgroup_id, ...)
+//     %store_index = affine.delinearize_index(%lane_id, ...)
+//     amdgpu.gather_to_lds %source[%thread_index] into %dest[%store_index]
+//   }
+struct LowerCoalescedGatherDMAPattern : public OpRewritePattern<scf::ForallOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LowerCoalescedGatherDMAPattern(MLIRContext *context,
+                                 ArrayRef<int64_t> workgroupSize,
+                                 int64_t subgroupSize,
+                                 int64_t maxLoadInstructionBits)
+      : OpRewritePattern<scf::ForallOp>(context), workgroupSize(workgroupSize),
+        subgroupSize(subgroupSize),
+        maxLoadInstructionBits(maxLoadInstructionBits) {}
+
+  LogicalResult matchAndRewrite(scf::ForallOp forallOp,
+                                PatternRewriter &rewriter) const override {
+    // Verify that the forall has the required GPU thread mapping.
+    if (failed(verifyThreadMapping(forallOp))) {
+      return failure();
+    }
+
+    // Ensure this is not nested inside another thread-mapped forall
+    if (forallOp->getParentOfType<scf::ForallOp>()) {
+      auto parentForall = forallOp->getParentOfType<scf::ForallOp>();
+      if (parentForall.getMapping()) {
+        auto parentMapping = *parentForall.getMapping();
+        if (!parentMapping.empty() &&
+            isa<gpu::GPUThreadMappingAttr>(parentMapping[0])) {
+          return rewriter.notifyMatchFailure(
+              forallOp, "nested thread-mapped forall not supported");
+        }
+      }
+    }
+
+    // Check that the forall body contains exactly one DMA op (plus terminator)
+    Block &body = forallOp.getRegion().front();
+    if (body.getOperations().size() != 2) {
+      return rewriter.notifyMatchFailure(
+          forallOp,
+          "forall body must have exactly 2 operations (dma + terminator)");
+    }
+
+    auto dmaOp = dyn_cast<IREE::GPU::CoalescedGatherDMAOp>(body.front());
+    if (!dmaOp) {
+      return rewriter.notifyMatchFailure(
+          forallOp,
+          "first operation in forall body is not a CoalescedGatherDMAOp");
+    }
+
+    // Verify memory layout
+    if (failed(verifyMemoryLayout(dmaOp, rewriter))) {
+      return failure();
+    }
+
+    Location loc = forallOp.getLoc();
+
+    // Set insertion point before the forall to create replacement ops
+    rewriter.setInsertionPoint(forallOp);
+
+    auto destMemRefType = dyn_cast<MemRefType>(dmaOp.getInit().getType());
+    int64_t numCopiesPerThread = getNumOfGlobalLoadsPerThread(
+        destMemRefType, subgroupSize, maxLoadInstructionBits);
+
+    Value indices = dmaOp.getIndices();
+    Value dest = dmaOp.getInit();
+    Value source = dmaOp.getSource();
+    auto indicesType = cast<MemRefType>(indices.getType());
+    auto destType = cast<MemRefType>(dest.getType());
+
+    int64_t numIndices = indicesType.getNumElements();
+
+    int64_t totalBytesOfCopy =
```

Comment:
I re-raise my note that there isn't a maximum here - that the DMA has multiple valid "these indices must be globally contiguous" variants